Update IV Objetivo 2 [IV-21-22]

eantoniocalo18 opened the pull request:
Entrega del objetivo 2 incluyendo:
- Elección del lenguaje de programación 
- Actualización de .gitignore para python
- Se incluyen #29 y #30 
- Se ha implementado las clases Horario y Proyecto y los problemas relacionados durante el proceso se han reflejado en #41 y #42

eantoniocalo18 committed:
Closes #28 Elección de lenguaje de programación Python

Actualizando .gitignore para el uso de python por su facilidad, legibilidad y escalabilidad.

eantoniocalo18 committed:
Avanza #41, Definición de una estructura para representar horas trabajadas

Se ha creado la clase Horario.py cuyos atributos son : id, usuario_relacionado,
horas_semanales y horas_trabajadas.
Con esta clase se pretende almacenar toda la información relacionada con las
horas trabajadas por un usuario

eantoniocalo18 committed:
Closes #30 y Avanza #41. Añadiendo método horas_restantes

Con la definición de este método se ha dado solución a #30 al permitir al usuario
consultar en cualquier momento las horas que le quedan por trabajar en la semana actual.
Además, se ha avanzado en M0

eantoniocalo18 committed:
Closes #29, Avanza #41 Añade método to_CSV()

Se ha creado un método to_CSV() que permitirá al usuario exportar los datos referentes
a su horario a dicho formato. Esto le permitirá ahorrar tiempo puesto que antes
tenía que hacerlo manualmente, solucionando así #29.
Además, avanza en M0

eantoniocalo18 committed:
Create vi.yaml

Creando vi.yaml para completar el hito del Objetivo 2

eantoniocalo18 committed:
Closes #42, definiendo una estructura para almacenar proyecto

Se ha creado la clase Proyecto.py con los atributos: id, planificacion y usuarios_asignados
en la que se almacenarán los datos referentes a las horas trabajadas por los usuarios y la planificacion

eantoniocalo18 committed:
Closes #42, creación de una estructura que almacena información de un proyecto

Se ha creado la clase Proyecto.py con los siguientes atributos:
- Id : Identificador único para cada proyecto
- Planificacion: Duración (en horas) estimada para cada proyecto
- Horarios : Vector de horarios donde estará incluida la información referente a las horas trabajadas por los diferentes horarios

eantoniocalo18 committed:
Fixing names

Arreglando los nombres de archivo propuestos por @morevi

morevi reviewed (changes requested):
- [x] Los nombres de los .py deben estar en minúscula.
- [x] vi.yaml -> iv.yaml
@eantoniocalo18

eantoniocalo18 commented:
@JJ Listo para revisión

JJ commented on src/horario.py:
```diff
@@ -0,0 +1,46 @@
+   
+    Publicos
+    id: str
+        uuid o id que identifica cada horario único
```
¿Por qué se tiene que identificar un horario de forma única? ¿En qué sirve eso a la lógica de negocio?

eantoniocalo18 replied:
Se ha presentado el problema en #48

JJ commented on src/horario.py:
```diff
@@ -0,0 +1,46 @@
+	self.horas_trabajadas = horas_trabajadas
+
+    def horas_restantes():
+	" Método que devolverá la diferencia entre horas_semanales y horas_trabajadas"
```
¿De dónde sale horas semanales?

JJ reviewed (changes requested):
Me da la impresión de que más allá de poner nombres razonables, no se está tratando de implementar en estructuras de datos diferentes un modelo del problema analizado en profundidad.

eantoniocalo18 committed:
Closes #30, se modifica el método get_horas_Trab

Se ha completado la información correspondiente al issue #30 y se ha visto la necesidad que se planteaba con este problema y también se ha visto que el atributo horas_semanales será un atributo que no está en la clase Horario

eantoniocalo18 committed:
Closes #48, cambia id por fecha_inicio

Se ha llegado a la conclusión de que tener como identificador la fecha de inicio es más rentable por varios motivos:
- Un usuario no puede trabajar dos veces en una misma semana, por lo tanto no se repetirá
- Es más orientativo que un id
- Como los horarios son todos de 7 días, permite saber el principio y fin de un horario 
Avanzando en M0

eantoniocalo18 commented:
Se ha actualizado la descripción de los issues y corregido los errores.
@JJ Listo para revisión.

JJ commented on src/horario.py:
```diff
@@ -0,0 +1,43 @@
+    def __init__(self, fecha_inicio, usuario_relacionado, horas_trabajadas):
+        self.fecha_inicio = fecha_inicio
+        self.nombre = usuario_relacionado
+	self.horas_trabajadas = horas_trabajadas
```
Por favor, intentad compilar lo que estáis haciendo.

JJ commented on src/proyecto.py:
```diff
@@ -0,0 +1,29 @@
+    horarios : Horario []
+        array con los horarios en los que se almacena la información de las horas trabajads por cada usuario
+    """
+    def __init__(self, id, planificacion, horarios):
```
A ver, no es que yo entienda mucho de esto. Pero ¿en un proyecto participa todo el mundo? ¿O sólo puedes poner horas en los proyectos que participas? ¿Hacemos por favor un intento de modelar de verdad el dominio del problema siguiendo domain driven design?

JJ reviewed (changes requested):
Este objetivo es fundamental, y recibe una buena cantidad de puntos en la nota final. Contiene una serie de conceptos fundamentales sobre diseño y programación. Un poco de esfuerzo por entender esos conceptos, que implican principalmente pasar de una serie de necesidades a estructuras de datos que cubran esas necesidades, es fundamental. Y eso por parte de todas las partes implicadas.

eantoniocalo18 committed:
Update proyecto.py añadiéndo método add_horas

Se ha modificado la clase proyecto teniendo en cuenta los comentarios hechos en #43 y en el issue #42.
Se ha comprobado la necesidad de establecer un atributo que almacene el usuario y las horas trabajadas por el mismo en lugar de almacenar los horarios creados por los usuarios.
Por otro lado, con esta nueva implementación de la clase proyecto, es necesario de un método que permita añadir las horas trabajadas por un usuario.
Se avanza en M0.

eantoniocalo18 committed:
Update horario.py fixing syntax errors

Basándome en los comentarios de #43 he corregido errores de sintaxis que habían surgido.

eantoniocalo18 commented:
Se han corregido los errores, ampliado el análisis de los issues #41 y #42 y también se ha profundizado en el funcionamiento que tendrá la clase proyecto.
@JJ Listo para revisión.

JJ reviewed (approved):
Podrías añadir un poco de validación a lo que pasas al constructor, para asegurarte que es lo qeu dice, pero si @morevi no te lo pide, no voy a ser yo quien lo haga...

morevi committed:
Merge branch 'main' into main
